libff/algebra/curves/profile/algebra_groups_read_profile.cpp

```diff
@@ -75,40 +90,647 @@ bool profile_group_read_sequential_uncompressed(const std::string &identifier)
+        if (0 <= ret) {
+            // std::cout << "cb_wait: done\n";
+            return ret;
+        }
+
+        if (ECANCELED == err) {
+            throw std::runtime_error("aio_error: cancelled");
+        }
+
+        throw std::runtime_error(
+            std::string("error from aio_error: ") + strerror(errno));
+    }
+}
+
+template<typename GroupT>
+void profile_group_read_random_aio_ordered_uncompressedw(
+    const std::string &identifier, const size_t interval)
+{
+    const std::string filename = get_filename(identifier);
+
+    // Treat the file as being divided into NUM_ELEMENTS_TO_READ groups, of
+    // size NUM_ELEMENTS_IN_FILE / NUM_ELEMENTS_TO_READ, reading from a ranadom
+    // offset within each group.
+
+    const size_t GROUP_SIZE = interval;
+
+    // Read from group_offset[i] in i-th group, i.e. global offset:
+    //   GROUP_SIZE * i + group_offset[i]
+    std::vector<size_t> group_offsets;
+    group_offsets.reserve(NUM_ELEMENTS_TO_READ);
+    for (size_t i = 0; i < NUM_ELEMENTS_TO_READ; ++i) {
+        group_offsets.push_back((size_t)rand() * GROUP_SIZE / RAND_MAX);
+    }
+
+    int fd = open(filename.c_str(), O_RDONLY);
+    if (fd < 0) {
+        throw std::runtime_error("failed to open " + filename);
+    }
+
+    aiocb cb1;
+    aiocb cb2;
+    GroupT dest1;
+    GroupT dest2;
+    const size_t size_on_disk = 2 * sizeof(dest1.X);
+
+    aiocb *cur_cb = &cb1;
+    aiocb *next_cb = &cb2;
+    GroupT *cur_dest = &dest1;
+    GroupT *next_dest = &dest2;
+
+    std::cout << "  Async Read '" << filename.c_str() << "' ("
+              << std::to_string(NUM_ELEMENTS_TO_READ) << " of "
+              << std::to_string(NUM_ELEMENTS_IN_FILE) << " elements ...\n";
+
+    {
+        enter_block("Read group elements profiling");
+
+        // Enqueue first request
+        cb_init(cur_cb, fd, group_offsets[0], size_on_disk, cur_dest);
+        cb_enqueue(cur_cb);
+
+        // Enqueue all requests
+        for (size_t i = 1; i < NUM_ELEMENTS_TO_READ; ++i) {
+            // Enqueue next (i-th)
+            const size_t file_offset =
+                (i * GROUP_SIZE + group_offsets[i]) * size_on_disk;
+            cb_init(next_cb, fd, file_offset, size_on_disk, next_dest);
+            cb_enqueue(next_cb);
+
+            // Wait for current and process
+            ssize_t bytes = cb_wait(cur_cb);
+            if (bytes != size_on_disk) {
+                throw std::runtime_error("insufficient bytes read");
+            }
+            // TODO: Check values
+
+            // Swap (at which point, cur_cb is the next element to wait for, and
+            // next_cb is unused).
+            std::swap(cur_cb, next_cb);
+            std::swap(cur_dest, next_dest);
+        }
+
+        // Wait for last request
+        cb_wait(cur_cb);
+        // TODO: Check values
+
+        leave_block("Read group elements profiling");
+    }
+}
+
+template<size_t BATCHSIZE> class batched_aio_reader
```

Comment:
As per my comment above, what about reading our group elements in "chunks", each chunk can be consumed with "batches". I think that would b good to avoid vocab collision when we can.

Reply:
`BATCHSIZE` should be PascalCase for consistency e.g. https://github.com/clearmatics/zeth/blob/master/libzeth/circuits/joinsplit.tcc#L23 or https://github.com/clearmatics/libff/blob/develop/libff/algebra/curves/curve_serialization.hpp#L22

Reply:
> As per my comment above, what about reading our group elements in "chunks", each chunk can be consumed with "batches". I think that would b good to avoid vocab collision when we can.

"batch" here is a genuine batch - a set of read operations performed at once.  So I think it's the correct term here.

> BATCHSIZE should be PascalCase for consistency ....

changed.

Reply:
> "batch" here is a genuine batch - a set of read operations performed at once. So I think it's the correct term here.

I'm not suggesting to rename `batch` here, rather I'm suggesting to rename `group` to `chunk` (batches of read will be performed on each chunk) -> the point being to avoid vocab collision with "group".